### ISSUE TITLE
Import function with array of ref<SomeClass> as parameter

### ISSUE THREAD

**Issue by Roms1383** [jac3km4/red4ext-rs#24]
Hi jekky, hope you're doing fine :)
Here's the report for the issue I mentioned to you on Discord.
Sorry for the weird naming, it's directly copied from my current mod dev.

### ✅ works with array of primitives

<details><summary>Redscript</summary>

```swift
module Addicted

public class Consumption extends IScriptable {
   public persistent let doses: array<Float>;
   public func SetDoses(value: array<Float>) -> Void { this.doses = value; }
}
```

</details>

<details><summary>Rust</summary>

```rs
#[derive(Default, Clone)]
#[repr(transparent)]
pub struct Consumption(pub Ref<IScriptable>);
unsafe impl RefRepr for Consumption {
    type Type = Strong;
    const CLASS_NAME: &'static str = "Addicted.Consumption";
}

#[redscript_import]
impl Consumption {
    fn set_doses(&mut self, values: Vec<f32>) -> ();
}
```

</details>

### ❌fails with array of e.g. `Consumption(Ref<IScriptable>)`

<details><summary>Redscript</summary>

```swift
module Addicted

public class Consumption extends IScriptable {
   public persistent let current: Int32;
}

public class Consumptions extends IScriptable {
   private persistent let values: array<ref<Consumption>>;
   public func SetValues(values: array<ref<Consumption>>) -> Void { this.values = values; }
}
```

</details>

<details><summary>Rust</summary>

```rs
#[derive(Default, Clone)]
#[repr(transparent)]
pub struct Consumption(Ref<IScriptable>);
unsafe impl RefRepr for Consumption {
    type Type = Strong;
    const CLASS_NAME: &'static str = "Addicted.Consumption";
}

#[derive(Default, Clone)]
#[repr(transparent)]
pub struct Consumptions(Ref<IScriptable>);
unsafe impl RefRepr for Consumptions {
    type Type = Strong;
    const CLASS_NAME: &'static str = "Addicted.Consumptions";
}

#[redscript_import]
impl Consumptions {
    fn set_values(&mut self, values: Vec<Consumption>) -> (); // or RedArray<Consumption>, same outcome
}
```

</details>

Error: `[2023-09-30 13:11:02.785] [addicted] [error] Function 'Addicted.OnIngestedItem' has panicked: failed to invoke SetValues;array<Addicted.Consumption>: function not found`

### ✅ works with array of `Ref<IScriptable>`

<details><summary>Redscript</summary>

```swift
public class Consumptions extends IScriptable {
    private persistent let values: array<ref<Consumption>>;
    // sadly the array has to be reconstructed
    public func SetValues(values: array<ref<IScriptable>>) -> Void {
        ArrayClear(this.values);
        this.values = [];
        for value in values {
            ArrayPush(this.values, value as Consumption);
        }
    }
}
```

</details>

<details><summary>Rust</summary>

```swift
#[derive(Default, Clone)]
#[repr(transparent)]
pub struct Consumption(pub Ref<IScriptable>);
unsafe impl RefRepr for Consumption {
    type Type = Strong;
    const CLASS_NAME: &'static str = "Addicted.Consumption";
}

#[derive(Default, Clone)]
#[repr(transparent)]
pub struct Consumptions(Ref<IScriptable>);
unsafe impl RefRepr for Consumptions {
    type Type = Strong;
    const CLASS_NAME: &'static str = "Addicted.Consumptions";
}

#[redscript_import]
impl Consumptions {
    fn set_values(&mut self, values: Vec<Ref<IScriptable>>) -> ();
}
```

</details>

---

The solution is a bit cumbersome, since it involves on Rust side:

```rs
let values = values.into_iter().map(|x| x.0).collect();
self.set_values(values);
```

Followed on Redscript side by:

```swift
ArrayClear(this.values);
this.values = [];
for value in values {
    ArrayPush(this.values, value as Consumption);
}
```

**Comment by Roms1383** [comment]
Ok I made a further test and can confirm, this also works and spare reconstructing the array on Redscript side:

```swift
public class Consumptions extends IScriptable {
    private persistent let values: array<ref<Consumption>>;
    public func SetValues(values: array<ref<Consumption>>) -> Void { this.values = values; }
}
```
```rs
#[derive(Default, Clone)]
#[repr(transparent)]
pub struct Consumptions(Ref<IScriptable>);
unsafe impl RefRepr for Consumptions {
    type Type = Strong;
    const CLASS_NAME: &'static str = "Addicted.Consumptions";
}

#[redscript_import]
impl Consumptions {
    fn set_values(&mut self, values: Vec<Ref<IScriptable>>) -> ();
}
```

This actually makes perfect sense since `Consumption` is `#[repr(transparent)]` over `Ref<IScriptable>` so `array<ref<Consumption>>` is akin to `array<ref<IScriptable>>` as far as the bridge is concerned.

If I have some time I'll try to come up with a PR, but not sure how to approach it: any suggestion is welcomed :)

**Comment by Roms1383** [comment]
**⚠️ Edit**: I think I was wrong with my latest comment.

I didn't log enough yesterday but when I tested this morning, on first consumption of `Items.FirstAidWhiffV0` (already in backpack) it seems ok (at least no error), but then as soon as I `Game.AddToInventory("Items.FirstAidWhiffVEpic", X)`, switched to it and consumed again I ran into the `SetValues;array<IScriptable> : function not found` again.

I'm currently reverting to using `Vec<Ref<IScriptable>>` / `array<ref<IScriptable>>` on both sides, which seems to work with the same scenario.
